test/deps/panic-checker

```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env python3
+
+import sys
+import re
+
+# Looks for a panic by grepping for stacktraces in a log file.
+# If a panic is found, it exits with code 1 and prints the stacktrace plus the last log message.
+# Only the first panic is returned.
+# If no panics are found it exits with code 0
+#
+# NOTE: When a panic occurs in LXD at runtime via a mux handler, it is logged at info level because the net/http library
+# has a built-in recover. We are not handling panic recovery manually. Because it is logged at info level, this checker
+# will only find panics if the test suite is run with LXD_VERBOSE=1 or LXD_DEBUG=1.
+with open(sys.argv[1]) as file:
+    output = ""
+    lastline = ""
+    for line in file:
+        if output == "" and not re.search("^goroutine\s+\d+\s+\[running]:", line):
+            # Output empty and no regex match so nothing found yet.
+            lastline = line
+            continue
+
+        if lastline != "":
+            # Got a regex match on this line, start the output with the last line and set it empty.
+            output = lastline + line
+            lastline = ""
+            continue
+
+        if re.search("(INFO|DEBUG|TRACE|WARNING|ERROR)", line):
+            # Print the stack trace and exit 1 when we encounter the next standard log message
+            sys.stdout.write(output)
+            exit(1)
+        else:
+            # While we don't have standard log messages, append the output (the stacktrace).
+            output += line
+
+# If we found a panic at the end of the log file, print it and exit 1.
+if output != "":
+    sys.stdout.write(output)
+    exit(1)
+
+# If we didn't find a panic, exit 0.
+exit(0)
+
+
```

Comment:
I couldn't help but test my feedback on a local copy. Since it's a small script I figured it was quicker to just provide it here:

```
#!/usr/bin/env python3

import sys
import re

# Looks for a panic by grepping for stacktraces in a log file.
# If a panic is found, print the last log message plus the stacktrace, exit 1
# If no panics are found, exit 0
#
# Only the first panic is returned.
#
# NOTE: When a panic occurs in LXD at runtime via a mux handler, it is logged
# at info level because the net/http library has a built-in recover. We are not
# handling panic recovery manually. Because it is logged at info level, this
# checker will only find panics if the test suite is run with LXD_VERBOSE=1 or
# LXD_DEBUG=1.
with open(sys.argv[1]) as file:
    found = False
    lastline = ""
    stacktrace_regex = re.compile(r'^goroutine\s+\d+\s+\[running\]:')
    standard_log_regex = re.compile(r'(INFO|DEBUG|TRACE|WARNING|ERROR)')

    for line in file:
        if not found and not stacktrace_regex.search(line):
            # Nothing found yet but lets retain the last log line
            lastline = line
            continue

        # Stacktrace detected, print the last log line that preceeded it
        if not found:
            sys.stderr.write(lastline)
            found = True

        # The first standard log message indicates the end of the stacktrace
        if standard_log_regex.search(line):
            break

        # Print the line as it is part of the stacktrace
        sys.stderr.write(line)

    if found:
        # Panic found, failure
        exit(1)

# No panic found, success
exit(0)
```

It's a bit of a rewrite (sorry) but I hope it reads better and is a tad simpler?

Some of the changes:

* Compile the regexps once, outside the loop
* Use `r""` string for regexps
* Don't accumulate output, just print it as it comes
* Use `stderr` instead of `stdout`

Reply:
Very neat improvements!

Also helps with the case where the panic log message is the last on the file.